### DISCUSSION TITLE
How does Wing fit with a serverless application?

### DISCUSSION THREAD

**Discussion by pgarbe** [winglang/wing#862] [General]
Wing is still elusive to me. I understood that there are so called resources which can be a real cloud resource (e.g. Bucket) or more abstract (e.g. Counter). With Elad I have clarified that cross-cutting concerns, like encryption are rather located at (cloud-) specific backends.

But where I struggle is how a serverless application would look like.  For example a web application to upload documents to a bucket. After the upload, the document is translated using a managed cloud service and stored in another bucket.

What could the integration look like with the frontend? The code for presigned URLs for upload might be very specific to the cloud vendor. How could I test this in the simulator?
Probably there is no “Translate” resource in Wing either (or is there?). A service like Amazon Translate is only available on AWS and not on other clouds (I guess). How would I implement this in Wing? Would it mean to write separate resources for each cloud service?

I know that Wing is in an early stage. But maybe you have some ideas how that could look like? It would help me to make the whole thing more tangible.

**Comment by ekeren** [comment]
Thanks @pgarbe, let me try to turn this into a code example

To implement the above example we need 2 types of resources: 
-  Bucket - The buckets are part of wing standard library (under `cloud.` namespace)  
-  Aws/Translate -  a wing package that is probably not part of the wing standard library, in the future installing such package might be via `npm install --save some_external_translate` command. 

To match @pgarbe example, lets assume that `some_external_translate` only implements the AWS target

Here is wing code 

```js (wing)
bring cloud;
bring some_external_translate; //This translate is only implemented for AWS

let origin_bucket = new cloud.Bucket() as "origin"
let translated_bucket = new cloud.Bucket() as "translated"
let translator = new some_external_translate.Translator();

origin.on_update(((key: str) ~> {
  let content = origin_bucket.get(key);
  let translated = translator.translate(content);
  translated_bucket.put(key, translated);
}))
```
Because `some_external_translate` package is not implemented for `sim` target I should get an error when I compile it to sim
```sh
wing compile <file>
ERROR: no "sim" target for some_external_translate.Translator
```

As the owner of this NPM package, not providing a sim solution is a very bad experience for the package consumers.
The owner can choose to create a simple and fake translator that just turns all lowercase letters to uppercase, if he does that 
the developer consuming this package can still see that his flow is working as expected in the sim (assuming he doesn't want to actual test the translate in his localhost environment).

This is quite simple, all that is needed to implement is a `some_external_translate.Translator` client for target `sim`.

Here is a naive implementation: 

```ts

export class Translator implements ITranslatorClient, ISimulatorResourceInstance {
  private readonly context: ISimulatorContext;

  public constructor(
     context: ISimulatorContext
  ) {
     this.context = context;
  }

  public async init(): Promise<void> {}
  public async cleanup(): Promise<void> {}

  public async translate(content: string): Promise<string> {
    return content.toLocaleUpperCase();
  }
}
```

Now the developer of this application can compile this only for aws and sim, but it might be good enough for the business need of the app.

Eventually there will be a couple of external `translate` packages, the best one will probably be the most complete solution that is implemented for the most targets in the most efficient way. People will choose the package based on downloads, forks,  just like the current npm eco system works today

@pgarbe does this help?

**Reply by ekeren** [reply]
I am actually really anticipating the world of possibilities that adding a custom resource will open  for us.


Will aim to have this prioritized on the todo app after that story is completed

**Reply by schosterbarak** [reply]
Would it be possible to debug across languages (js and winglang?)

**Reply by ekeren** [reply]
We are planning to allow winglang apps to `bring` javascript npm packages, which means that in practice js and winglang will run side by side under the same runtime.  

So it seems relevant for us to prioritize a user story that cover this use case, something like:
> "As a developer I would like to debug my winglang application including the javascript npm packeage that it `bring`"

In the example of the code above, I actually think that it will be best to implement the above resource using winglang and not typescript/javascript but this is not available for the time being

**Comment by pgarbe** [comment]
Thanks @ekeren for that detailed explanation. That makes sense to me. I guess the challenge comes with deciding what is part of the core library and what not. What's your idea or process on defining what should be part of the core library? I've seen Endpoint in the backlog. Straight forward from the perspective of an application developer. But I'm curious about the implementation. There are many possibilities how that could translate into AWS.

**Reply by ekeren** [reply]
Completely agree that this is a real challenge, @Chriscbr is working on the initial SDK specification (#858) , you can take a look at the initial proposal and chatter on the PR.  

These are early days, for the time being we are laying some guidelines and initial state and look for feedback to improve. Eventually if we are able to create a strong eco-system, I believe that the best ideas and common practices will emerge from the developers using wing and the owners of custom resources. 

If you have any thoughts of a process we should take, take a look at this discussion covering wing eco-system (#652), I think this is probably the a good place to get this discussion going.
